Mathlib/MeasureTheory/Decomposition/Exhaustion.lean

```diff
@@ -0,0 +1,335 @@
+  split_ifs with h
+  · exact h.choose_spec.1
+  · exact MeasurableSet.empty
+
+instance : SigmaFinite (μ.restrict (μ.sigmaFiniteSetWRT ν)) := by
+  rw [Measure.sigmaFiniteSetWRT]
+  split_ifs with h
+  · exact h.choose_spec.2.1
+  · rw [Measure.restrict_empty]
+    infer_instance
+
+section IsFiniteMeasure
+
+variable [IsFiniteMeasure ν]
+
+/-! We prove that the condition in the definition of `sigmaFiniteSetWRT` is true for finite
+measures. Since every s-finite measure is absolutely continuous with respect to a finite measure,
+the condition will then also be true for s-finite measures. -/
+
+/-- Let `C` be the supremum of `ν s` over all measurable sets `s` such that `μ.restrict s` is
+sigma-finite. `C` is finite since `ν` is a finite measure. Then there exists a measurable set `t`
+with `μ.restrict t` sigma-finite such that `ν t ≥ C - 1/n`. -/
+lemma exists_isSigmaFiniteSet_measure_ge (μ ν : Measure α) [IsFiniteMeasure ν] (n : ℕ) :
+    ∃ t, MeasurableSet t ∧ SigmaFinite (μ.restrict t)
+      ∧ (⨆ (s) (_ : MeasurableSet s) (_ : SigmaFinite (μ.restrict s)), ν s) - 1/n ≤ ν t := by
+  by_cases hC_lt : 1/n < ⨆ (s) (_ : MeasurableSet s) (_ : SigmaFinite (μ.restrict s)), ν s
+  · have h_lt_top : ⨆ (s) (_ : MeasurableSet s) (_ : SigmaFinite (μ.restrict s)), ν s < ∞ := by
+      refine (?_ : ⨆ (s) (_ : MeasurableSet s)
+        (_ : SigmaFinite (μ.restrict s)), ν s ≤ ν Set.univ).trans_lt (measure_lt_top _ _)
+      refine iSup_le (fun s ↦ ?_)
+      exact iSup_le (fun _ ↦ iSup_le (fun _ ↦ measure_mono (Set.subset_univ s)))
+    obtain ⟨t, ht⟩ := exists_lt_of_lt_ciSup
+      (ENNReal.sub_lt_self h_lt_top.ne (ne_zero_of_lt hC_lt) (by simp) :
+          (⨆ (s) (_ : MeasurableSet s) (_ : SigmaFinite (μ.restrict s)), ν s) - 1/n
+        < ⨆ (s) (_ : MeasurableSet s) (_ : SigmaFinite (μ.restrict s)), ν s)
+    have ht_meas : MeasurableSet t := by
+      by_contra h_not_mem
+      simp only [h_not_mem] at ht
+      simp at ht
+    have ht_mem : SigmaFinite (μ.restrict t) := by
+      by_contra h_not_mem
+      simp only [h_not_mem] at ht
+      simp at ht
+    refine ⟨t, ht_meas, ht_mem, ?_⟩
+    simp only [ht_meas, ht_mem, iSup_true] at ht
+    exact ht.le
+  · refine ⟨∅, MeasurableSet.empty, by rw [Measure.restrict_empty]; infer_instance, ?_⟩
+    rw [tsub_eq_zero_of_le (not_lt.mp hC_lt)]
+    exact zero_le'
+
+/-- A measurable set such that `μ.restrict (μ.sigmaFiniteSetGE ν n)` is sigma-finite and
+for `C` the supremum of `ν s` over all measurable sets `s` with `μ.restrict s` sigma-finite,
+`ν (μ.sigmaFiniteSetGE ν n) ≥ C - 1/n`. -/
+def Measure.sigmaFiniteSetGE (μ ν : Measure α) [IsFiniteMeasure ν] (n : ℕ) : Set α :=
+  (exists_isSigmaFiniteSet_measure_ge μ ν n).choose
+
+lemma measurableSet_sigmaFiniteSetGE (μ ν : Measure α) [IsFiniteMeasure ν] (n : ℕ) :
+    MeasurableSet (μ.sigmaFiniteSetGE ν n) :=
+  (exists_isSigmaFiniteSet_measure_ge μ ν n).choose_spec.1
+
+instance (n : ℕ) : SigmaFinite (μ.restrict (μ.sigmaFiniteSetGE ν n)) :=
```

Comment:
Since this construction is purely local to this file, I wouldn't register it as a global instance, to avoid cluttering instance search. Can you make it a local instance instead?